sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/SymbolicExecution/ExplodedGraphTests.cs

```diff
@@ -538,22 +537,58 @@ public int Main(Dictionary<string, string> dictionary, string key)
+
+                            args.ProgramState.GetSymbolValue(valueSymbol).Should().NotBeNull();
+                            valueSymbol.HasConstraint(ObjectConstraint.NotNull, args.ProgramState).Should().BeTrue();
                             break;
 
                         case "dictionary.TryGetValue(key, out var value)":
                             args.ProgramState.GetSymbolValue(dictionarySymbol).Should().NotBeNull();
                             dictionarySymbol.HasConstraint(ObjectConstraint.NotNull, args.ProgramState).Should().BeTrue();
 
-                            // Currently the DeclarationExpressionSyntax are ignored so the "value" variable is not
-                            // https://github.com/SonarSource/sonar-dotnet/issues/2936
-                            args.ProgramState.GetSymbolValue(valueSymbol).Should().BeNull();
+                            args.ProgramState.GetSymbolValue(valueSymbol).Should().NotBeNull();
+                            valueSymbol.HasConstraint(ObjectConstraint.NotNull, args.ProgramState).Should().BeTrue();
+
+                            args.ProgramState.HasValue.Should().BeFalse();
                             break;
                     }
                 };
 
             context.WalkWithInstructions(5);
         }
 
+        [TestMethod]
+        [TestCategory("Symbolic execution")]
+        public void ExplodedGraph_DeclarationExpressionVisit_AsOutParameter_Discard()
+        {
+            const string testInput = "string key = null; Dictionary<string, int> dictionary = new Dictionary<string, int>(); dictionary.TryGetValue(key, out var _); dictionary.TryGetValue(key, out _);";
+            var context = new ExplodedGraphContext(testInput);
+            var dictionarySymbol = context.SemanticModel.GetDeclaredSymbol(context.MainMethod.ParameterList.Parameters.First());
+            context.ExplodedGraph.InstructionProcessed +=
+                (sender, args) =>
+                {
+                    var instruction = args.Instruction.ToString();
+                    switch (instruction)
+                    {
+                        case "dictionary.TryGetValue(key, out var _)":
+                            args.ProgramState.GetSymbolValue(dictionarySymbol).Should().NotBeNull();
+                            dictionarySymbol.HasConstraint(ObjectConstraint.NotNull, args.ProgramState).Should().BeTrue();
+
+                            args.ProgramState.HasValue.Should().BeFalse();
+                            break;
+
+                        case "dictionary.TryGetValue(key, out _)":
+                            args.ProgramState.GetSymbolValue(dictionarySymbol).Should().NotBeNull();
+                            dictionarySymbol.HasConstraint(ObjectConstraint.NotNull, args.ProgramState).Should().BeTrue();
+
+                            args.ProgramState.HasValue.Should().BeFalse();
```

Comment:
Before checking that the program state has empty stack here, it would be nice to check that it has value on the stack after having evaluated the `key` parameter